tests/custom-string-parsers.rs

```diff
@@ -0,0 +1,125 @@
+use std::ffi::{OsStr, OsString};
+
+#[derive(StructOpt, PartialEq, Debug)]
+struct PathOpt {
+    #[structopt(short = "p", long = "path", parse(from_os_str))]
+    path: PathBuf,
+
+    #[structopt(short = "d", default_value = "../", parse(from_os_str))]
+    default_path: PathBuf,
+
+    #[structopt(short = "v", parse(from_os_str))]
+    vector_path: Vec<PathBuf>,
+
+    #[structopt(short = "o", parse(from_os_str))]
+    option_path_1: Option<PathBuf>,
+
+    #[structopt(short = "q", parse(from_os_str))]
+    option_path_2: Option<PathBuf>,
+}
+
+#[test]
+fn test_path_opt_simple() {
+    assert_eq!(
+        PathOpt {
+            path: PathBuf::from("/usr/bin"),
+            default_path: PathBuf::from("../"),
+            vector_path: vec![
+                PathBuf::from("/a/b/c"),
+                PathBuf::from("/d/e/f"),
+                PathBuf::from("/g/h/i"),
+            ],
+            option_path_1: None,
+            option_path_2: Some(PathBuf::from("j.zip"))
+        },
+        PathOpt::from_clap(PathOpt::clap().get_matches_from(&[
+            "test",
+            "-p", "/usr/bin",
+            "-v", "/a/b/c",
+            "-v", "/d/e/f",
+            "-v", "/g/h/i",
+            "-q", "j.zip",
+        ]))
+    );
+}
+
+
+
+
+fn parse_hex(input: &str) -> Result<u64, ParseIntError> {
+    u64::from_str_radix(input, 16)
+}
+
+#[derive(StructOpt, PartialEq, Debug)]
+struct HexOpt {
+    #[structopt(short = "n", parse(try_from_str = "parse_hex"))]
+    number: u64,
+}
+
+#[test]
+fn test_parse_hex() {
+    assert_eq!(
+        HexOpt { number: 5 },
+        HexOpt::from_clap(HexOpt::clap().get_matches_from(&["test", "-n", "5"]))
+    );
+    assert_eq!(
+        HexOpt { number: 0xabcdef },
+        HexOpt::from_clap(HexOpt::clap().get_matches_from(&["test", "-n", "abcdef"]))
+    );
+
+    let err = HexOpt::clap().get_matches_from_safe(&["test", "-n", "gg"]).unwrap_err();
+    assert!(err.message.contains("invalid digit found in string"), err);
+}
+
+
+
+
+fn custom_parser_1(_: &str) -> &'static str {
+    "A"
+}
+fn custom_parser_2(_: &str) -> Result<&'static str, u32> {
+    Ok("B")
+}
+fn custom_parser_3(_: &OsStr) -> &'static str {
+    "C"
+}
+fn custom_parser_4(_: &OsStr) -> Result<&'static str, OsString> {
+    Ok("D")
+}
+
+#[derive(StructOpt, PartialEq, Debug)]
+struct NoOpOpt {
+    #[structopt(short = "a", parse(from_str = "custom_parser_1"))]
+    a: &'static str,
+    #[structopt(short = "b", parse(try_from_str = "custom_parser_2"))]
+    b: &'static str,
+    #[structopt(short = "c", parse(from_os_str = "custom_parser_3"))]
+    c: &'static str,
+    #[structopt(short = "d", parse(try_from_os_str = "custom_parser_4"))]
+    d: &'static str,
+}
+
+#[test]
+fn test_every_custom_parser() {
+    assert_eq!(
+        NoOpOpt { a: "A", b: "B", c: "C", d: "D" },
+        NoOpOpt::from_clap(NoOpOpt::clap().get_matches_from(&[
+            "test", "-a=?", "-b=?", "-c=?", "-d=?",
+        ]))
+    );
+}
```

Comment:
Lack tests with default value:
  * [ ] `parse(from_str)` (with a `Vec<u8>` for example)
  * [ ] `parse(try_from_str)` (with a `u64` for example)
  * [x] `parse(from_os_str)` (PathBuf)

and an empty line at the end of file

Reply:
`Vec<u8>` as a field doesn't work, because `Vec` and `Option` are still treated specially. I've defined a type alias instead.

BTW the following also work, but I think this is a defect:

```rust
#[derive(StructOpt, PartialEq, Debug)]
struct DefaultedOpt {
    #[structopt(short = "b", parse(from_str))]
    bytes: (Vec<u8>),    // <-- with '(' ... ')'.
}
```